docs/modules/serialization/pages/schema-evolution-full.adoc

```diff
@@ -0,0 +1,779 @@
+TYPE IMap
+OPTIONS (
+  'keyFormat' = 'compact',
+  'keyCompactTypeName' = 'com.example.order.v3.OrderKeyV3',
+  'valueFormat' = 'compact',
+  'valueCompactTypeName' = 'com.example.order.v3.OrderV3'
+)
+COLUMNS (
+  accountId BIGINT EXTERNAL NAME "__key.accountId",
+  id BIGINT EXTERNAL NAME "__key.id",
+  amount DECIMAL EXTERNAL NAME "this.amount",
+  status VARCHAR EXTERNAL NAME "this.status",
+  currency VARCHAR EXTERNAL NAME "this.currency"
+);
+----
+
+Old queries continue to work on the old map (`orders`), and new queries can target the new map (`orders_v3`).
+
+[source,sql]
+----
+-- Old schema
+SELECT id, customerId, amount FROM orders;
+
+-- New schema
+SELECT id, accountId, amount FROM orders_v3;
+----
+
+Each versioned map has its own mapping and query surface, so schema incompatibility never breaks existing queries.
+
+=== Using Views for a stable SQL layer
+
+If you want to expose a single SQL interface while migrating between versions, create a VIEW that merges both schemas:
+
+[source,sql]
+----
+CREATE OR REPLACE VIEW orders_latest AS
+SELECT id, customerId, amount, status, currency FROM orders
+UNION ALL
+SELECT id, accountId AS customerId, amount, status, currency FROM orders_v3;
+----
+
+Applications can query `orders_latest` without needing to know which schema version is active.
+
+== Index evolution
+
+Schema changes can affect how indexes are defined and used.
+This section explains when you can keep existing indexes, when you must recreate them, and how to handle indexes during migrations.
+
+=== Compatible schema changes
+
+Additive or type-widening changes (for example, adding `currency` or widening `int` → `bigint`) do not invalidate existing indexes on unchanged fields.
+
+* Indexes on old fields continue to work.
+* You may add new indexes for newly introduced fields at any time.
+
+=== Incompatible schema changes
+
+Backward-incompatible changes (field rename, removal, type narrowing, or new key structure) require new index definitions aligned with the new schema.
+
+* **Field rename/removal:** drop the old index and create a new one on the new field.
+* **Type narrowing:** avoid in place; migrate to a new map and define indexes there.
+* **Partitioning/key change (e.g. composite key):** create indexes on the **new map** only.
+
+[NOTE]
+====
+Indexes are scoped per map. Versioned maps (for example, `orders` and `orders_v3`) hold independent index metadata; define indexes separately for each.
+Create these indexes on the **target map first** (so queries are efficient as data arrives).
+====
+
+== Caveats and warnings
+
+* Certain field kinds (e.g. ARRAY_OF_COMPACT, COMPACT) have restricted evolution semantics (cannot change element kind or switch between nullable/non-nullable).
+* Default handling in Compact serialization:
+  * When a newer schema adds fields, Hazelcast automatically assigns type-specific defaults (`0`, `false`, or `null) when reading older data that does not include those fields.
```

Comment:
> Hazelcast automatically assigns type-specific defaults

this applies only to zero-config compact serialization, and then default value defined in the Java class is used. In explicitly implemeted readers, methods like `getInt32` throw if the field is not defined or has different type. the only elasticity they have is that they also handle nullable variants (by requiring that the value is provided).